### ISSUE TITLE
ConfigureSmartEnum Fails For OwnedEntity

### ISSUE THREAD

**Issue by mgraf1** [ardalis/SmartEnum#253]
Relevant project: Ardalis.SmartEnum.EFCore 2.1.0

When an object that is configured in EF Core as an owned entity contains a property that is SmartEnum. The call to `modelBuild.ConfigureSmartEnum()` fails with the following exception

```
The entity type '<Name of owned entity>' cannot be configured as non-owned because it has already been configured as a owned. Use the nested builder in `OwnsOne` or `OwnsMany` on the owner entity type builder to further configure this type. If you want to override previous configuration first remove the entity type from the model by calling 'Ignore'. See https://aka.ms/efcore-docs-owned for more information.'
```

Associated Ef Core Issue: https://github.com/dotnet/efcore/issues/27788

**Comment by Khitiara** [comment]
Transposing my comment from #221: This happens due to the use of `.Entity` referring to an owned entity type; a possible solution could be to use `ConfigureConventions` where you can use [`ModelConfigurationBuilder.Properties`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.modelconfigurationbuilder.properties?view=efcore-6.0#microsoft-entityframeworkcore-modelconfigurationbuilder-properties(system-type)) to specify a default conversion for a given type, though this has the downside of requiring either specifying conversions for all `SmartEnum`-derived types or registering all used smartenums individually

**Comment by mgraf1** [comment]
@Khitiara - We may be able to get around the downsides you mention through the use of reflection

```csharp
public static void ConfigureSmartEnum(this ModelConfigurationBuilder configurationBuilder)
{
    var modelBuilder = configurationBuilder.CreateModelBuilder(null);
    var propertyTypes = modelBuilder.Model.GetEntityTypes()
        .SelectMany(e => e.ClrType.GetProperties())
        .Where(p => TypeUtil.IsDerived(p.PropertyType, typeof(SmartEnum<,>)))
        .Select(p => p.PropertyType)
        .Distinct();

    foreach (var propertyType in propertyTypes)
    {
        var keyType = TypeUtil.GetValueType(propertyType, typeof(SmartEnum<,>));
        var converterType = typeof(SmartEnumConverter<,>).MakeGenericType(propertyType, keyType);

        configurationBuilder.Properties(propertyType)
            .HaveConversion(converterType);
    }
}
```

**Comment by mgraf1** [comment]
Hey @ardalis, what needs to happen to get a new version of the package with these changes? It looks like the build pipeline failed when #296 was merged in. Thanks!

**Comment by jeffward01** [comment]
> ```cs
>         var converterType = typeof(SmartEnumConverter<,>).MakeGenericType(propertyType, keyType);
> ```

This for some reason is not working for me.

I also extracted the code out that was missing from the release, and attempted to configure SmartEnum in the `ConfigureConventions` the code 'ran' but it still threw the `no CLR type was found` errror.

I then adapted your code to run in the modelbuilder:

```csharp
    private void ConfigureSmartEnum(ModelBuilder modelBuilder)
    {
        foreach (IMutableEntityType smartEnumType in modelBuilder.Model.GetEntityTypes()
                     .Where(p => SmartEnumTypeUtil.IsDerived(p.ClrType, typeof(SmartEnum<,>))))
        {
            modelBuilder.Entity(smartEnumType.ClrType)
                .HasKey("Value");
        }

        foreach (PropertyInfo smartEnumPropertyOnType in modelBuilder.Model.GetEntityTypes()
                     .SelectMany(e => e.ClrType.GetProperties())
                     .Where(p => SmartEnumTypeUtil.IsDerived(p.PropertyType, typeof(SmartEnum<,>))))
        {
            Type converterType = typeof(SmartEnumConverter<,>).MakeGenericType(
                smartEnumPropertyOnType.PropertyType,
                typeof(int));

            modelBuilder.Entity(smartEnumPropertyOnType.DeclaringType!)
                .Property(smartEnumPropertyOnType.Name)
                .HasConversion(converterType);
        }
 
    }
```

And have had no luck at getting this to work.

Also, the lack of the parameterless  constructor which was removed here: https://github.com/ardalis/SmartEnum/issues/60

is also problematic, I have to use a workaround like: 

```csharp


public abstract class IdentityUserRoleType : SmartEnum<IdentityUserRoleType, int>
{
    /// <summary>
    ///     The user role type
    /// </summary>
    public static readonly IdentityUserRoleType User = new UserRoleType();

    /// <summary>
    ///     The admin role type
    /// </summary>
    public static readonly IdentityUserRoleType Admin = new AdminRoleType();

    /// <summary>
    ///     The super admin role type
    /// </summary>
    public static readonly IdentityUserRoleType SuperAdmin = new SuperAdminRoleType();

    /// <summary>
    ///     A general role type with the value of 'AppIdentityUser'
    /// </summary>
    public static readonly IdentityUserRoleType General = new GeneralRoleType();

   // This is the work around
    protected IdentityUserRoleType() : base(string.Empty,0) { }
   // ^^^ This is the work around

    /// <summary>
    ///     Initializes a new instance of the <see cref="IdentityUserRoleType.RoleType" /> class
    /// </summary>
    /// <param name="name">The name</param>
    /// <param name="value">The value</param>
    private IdentityUserRoleType(
        string name,
        int value)
        : base(name, value)
    {
    }

// ... more code....
}
```

I know am removing all my owned types from my project and accepting defeat until this is further resolved 😢

**Comment by mgraf1** [comment]
@jeffward01 

Do you have a minimal example of a model with an owned entity that is not working for you? I could try to take a look.

The approach I finally landed on in PR #296, was a little different than what was mentioned in the comment above. The `ModelBuilder` approach was significantly different and required walking the property relationship graph and special handling for things like OwnsOne vs OwnsMany relationships. It can be found here:  https://github.com/ardalis/SmartEnum/pull/296/files#diff-a9419b753f712505ff3d8ac2ad943ae567a7d0f868d12dd97865d8090ea1d14f

**Comment by jeffward01** [comment]
@mgraf1 - big thanks for this, I appreciate the help! SmartEnum is one of my absolute favorite and most used libraries.

Currently, I implement my `SmartEnum` side by side, like this:

![image](https://user-images.githubusercontent.com/12589359/211743454-875b4673-28ef-429a-91cb-41ab672c79d2.png)

This is a simple example of my SmartEnum use, usually I have many more properties and such.

I am working on a big project, and wanted to merge the best of `EfCore` and `SmartEnum`, then I hit the road blocks and am here.

-----

I will create a small sample project of the issue in a few days, I'm currently on a deadline on this other project and won't have time to dive-deep into this for at least a few days. I wanted to let you know because this is something im very much interested in solving.

Thanks for the code, you made the complexities of the Entity Graph look simple! It's very easy to read your code (Usually when people work with the EfGraph, it gets a bit messy, quickly).

I'll give it a go and report back!

Thanks again for getting back to me! Hope your doing well

--------

On an absolutely different topic, here is an interesting library that works with the Entity Graph: https://github.com/leonardoporro/Detached-Mapper -- It's definitely not applicable to this issue we are talking about, or `SmartEnum` that I know of -- however its an interesting library that you might find interesting as well.

Thanks again
